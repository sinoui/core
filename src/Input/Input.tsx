/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { useState, useCallback } from 'react';
import BaseInput, { BaseInputProps } from '@sinoui/core/BaseInput';
import styled, { css } from 'styled-components';
import bemClassNames from '../utils/bemClassNames';

export interface InputProps extends BaseInputProps {
  error?: boolean;
  warning?: boolean;
  focused?: boolean;
  dense?: boolean;
}

const FULL_SCALE_X = 1;
const FULL_SCALE_Y = 1;

const errorOrWarnigStyle = css<InputProps>`
  && {
    &::after {
      background-color: ${(props) =>
        props.theme.palette[props.error ? 'error' : 'warning'][500]};
      transform: scaleX(${FULL_SCALE_X}) scaleY(${FULL_SCALE_Y});
    }
  }
`;

const transformStyle = (props: { focused?: boolean }) => {
  const scaleX = props.focused ? FULL_SCALE_X : 0;
  const scaleY = props.focused ? FULL_SCALE_Y : 0;
  return css`
    transform: scaleX(${scaleX}) scaleY(${scaleY});
  `;
};

const inkbarStyle = css`
  &::after {
    background-color: ${(props) =>
      props.theme.palette.primary[
        props.theme.palette.type === 'light' ? 700 : 300
      ]};
    left: 0;
    bottom: 0;
    content: '';
    height: 2px;
    position: absolute;
    right: 0;
    transition: ${(props) =>
      props.theme.transitions.create('transform', {
        duration: props.theme.transitions.duration.shorter,
        easing: props.theme.transitions.easing.easeOut,
      })};
    pointer-events: none;
    ${transformStyle};
  }
`;

const underlineStyle = css<InputProps>`
  &::before {
    background-color: ${(props) => props.theme.palette.input.bottomLine};
    left: 0;
    bottom: 0;
    content: '';
    display: block;
    height: 1px;
    position: absolute;
    right: 0;
    transition: ${(props) =>
      props.theme.transitions.create('background-color', {
        duration: props.theme.transitions.duration.shorter,
        easing: props.theme.transitions.easing.easeInOut,
      })};
    pointer-events: none;
    transform: scaleX(${FULL_SCALE_X}) scaleY(${FULL_SCALE_Y});
  }

  &:hover::before {
    ${(props) =>
      !props.disabled &&
      !props.readOnly &&
      !props.error &&
      `background-color: ${props.theme.palette.text.primary};
  height: 2px;`};
  }
`;

const disabledUnderlineStyle = css`
  &::before {
    background: transparent;
    background-image: linear-gradient(
      to right,
      ${(props) => props.theme.palette.input.bottomLine} 33%,
      transparent 0%
    );
    background-position: left top;
    background-repeat: repeat-x;
    background-size: 5px 1px;
  }
`;

const StyledBaseInput = styled(BaseInput)<{
  error?: boolean;
  warning?: boolean;
  dense?: boolean;
}>`
  color: ${(props) =>
    props.disabled
      ? props.theme.palette.text.disabled
      : props.theme.palette.input.inputText};
  ${(props) => (props.error || props.warning) && errorOrWarnigStyle};
  ${(props) => !props.disabled && inkbarStyle};
  ${(props) => !props.error && underlineStyle};
  ${(props) => props.disabled && disabledUnderlineStyle};

  > input {
    ${(props) => props.dense && `padding-top:3px;`}
  }

  label ~ & {
    margin-top: 16px;
  }

  &::before,
  &::after {
    position: absolute;
    bottom: 0px;
    /*
     * 修复缺陷：输入框的下划线不出现
     *
     * 原因：IE对于小高度的元素不会正常展示。参考
     * [Inability to Have Elements with Small Heights][link1]
     *
     * * [link1]: https://code.tutsplus.com/tutorials/9-most-common-ie-bugs-and-how-to-fix-them--net-7764
     */
    overflow: hidden;
    font-size: 0px;
  }
`;

/**
 * 输入框
 */
export default React.forwardRef<HTMLDivElement, InputProps>(function Input(
  props: InputProps,
  ref,
) {
  const {
    fullWidth = false,
    inputComponent = 'input',
    multiline = false,
    type = 'text',
    className,
    error,
    warning,
    disabled,
    readOnly,
    dense,
    onFocus,
    onBlur,
    ...other
  } = props;
  const [focused, setFocused] = useState(false);

  const handleBlur = useCallback(
    (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      if (!props.readOnly) {
        setFocused(false);
        if (onBlur) {
          onBlur(event);
        }
      }
    },
    [onBlur, props.readOnly],
  );

  const handleFocus = useCallback(
    (event: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      if (!props.readOnly) {
        setFocused(true);
        if (onFocus) {
          onFocus(event);
        }
      }
    },
    [onFocus, props.readOnly],
  );

  return (
    <StyledBaseInput
      fullWidth={fullWidth}
      error={error}
      warning={warning}
      focused={focused}
      inputComponent={inputComponent}
      multiline={multiline}
      type={type}
      ref={ref}
      disabled={disabled}
      readOnly={readOnly}
      dense={dense}
      {...other}
      className={bemClassNames(
        'sinoui-input',
        {
          disabled,
          readOnly,
          error,
          focused,
          dense,
        },
        className,
      )}
      onFocus={handleFocus}
      onBlur={handleBlur}
    />
  );
});
