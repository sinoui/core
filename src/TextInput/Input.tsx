import BaseInput from '@sinoui/core/BaseInput';
import type { BaseInputProps } from '@sinoui/core/BaseInput';
import styled, { css } from 'styled-components';
import lineRippleStyle from './lineRippleStyle';

export interface InputProps extends BaseInputProps {
  /**
   * 设置为`true`，则表示输入框处于聚焦状态。
   */
  focused?: boolean;
  /**
   * 设置为`true`，则表示输入框处于密集模式。
   */
  dense?: boolean;
  /**
   * 无标签
   */
  noLabel?: boolean;
}

const noLabelStyle = css`
  padding: 7.5px 0 8.5px;
`;

const denseNoLabelStyle = css`
  padding: 3.5px 0px 4.5px;
`;

/**
 * 标准文本输入框
 */
const StandardInput: React.SFC<InputProps> = styled(BaseInput)<InputProps>`
  color: ${({ theme, disabled }) =>
    disabled ? theme.palette.text.disabled : theme.palette.text.primary};

  label ~ & {
    margin-top: 16px;
  }

  .sinoui-form-item--floating > .sinoui-form-item__content & {
    margin-top: 16px;
  }

  ${({ multiline }) => (multiline ? '&' : '> .sinoui-base-input__input')} {
    padding: 3.5px 0px 4.5px;
    ${({ noLabel }) => noLabel && noLabelStyle}
    ${({ noLabel, dense }) => dense && noLabel && denseNoLabelStyle}
  }

  ${lineRippleStyle}
`;

export default StandardInput;
