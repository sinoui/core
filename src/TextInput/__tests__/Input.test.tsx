import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import Input from '../Input';

it('正确渲染input', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <Input />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染不可用的Input', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <Input disabled />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染错误状态下的input', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <Input error />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('渲染密集模式下的Input', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <Input dense />
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
