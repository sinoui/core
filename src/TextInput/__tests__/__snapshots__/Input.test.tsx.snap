// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正确渲染input 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  line-height: 1.1875em;
}

.c0 > input,
.c0 > textarea {
  padding: 6px 0 7px;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.1876em;
  min-height: 1.1876em;
}

.c0 > input::-moz-placeholder,
.c0 > textarea::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder,
.c0 > textarea::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder,
.c0 > textarea:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear,
.c0 > textarea::-ms-clear {
  display: none;
}

.c0 > input:focus,
.c0 > textarea:focus {
  outline: 0;
}

.c0 > input:invalid,
.c0 > textarea:invalid {
  box-shadow: none;
}

.c0 > textarea {
  resize: none;
  overflow: hidden;
}

.c1 {
  color: rgba(0,0,0,0.87);
}

.c1::after {
  background-color: #3f51b5;
  left: 0;
  bottom: 0;
  content: '';
  height: 2px;
  position: absolute;
  right: 0;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  -webkit-transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c1::before {
  background-color: rgba(0,0,0,0.42);
  left: 0;
  bottom: 0;
  content: '';
  display: block;
  height: 1px;
  position: absolute;
  right: 0;
  -webkit-transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
}

.c1:hover::before {
  background-color: rgba(0,0,0,0.87);
  height: 2px;
}

label ~ .c1 {
  margin-top: 16px;
}

.c1::before,
.c1::after {
  position: absolute;
  bottom: 0px;
  overflow: hidden;
  font-size: 0px;
}

<div
  className="c0 sinoui-base-input__layout c1 sinoui-input"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`渲染不可用的Input 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  color: rgba(0,0,0,0.38);
  font-size: 1rem;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  line-height: 1.1875em;
}

.c0 > input,
.c0 > textarea {
  padding: 6px 0 7px;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.1876em;
  min-height: 1.1876em;
}

.c0 > input::-moz-placeholder,
.c0 > textarea::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder,
.c0 > textarea::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder,
.c0 > textarea:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear,
.c0 > textarea::-ms-clear {
  display: none;
}

.c0 > input:focus,
.c0 > textarea:focus {
  outline: 0;
}

.c0 > input:invalid,
.c0 > textarea:invalid {
  box-shadow: none;
}

.c0 > textarea {
  resize: none;
  overflow: hidden;
}

.c1 {
  color: rgba(0,0,0,0.38);
}

.c1::before {
  background-color: rgba(0,0,0,0.42);
  left: 0;
  bottom: 0;
  content: '';
  display: block;
  height: 1px;
  position: absolute;
  right: 0;
  -webkit-transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
}

.c1::before {
  background: transparent;
  background-image: linear-gradient( to right, rgba(0,0,0,0.42) 33%, transparent 0% );
  background-position: left top;
  background-repeat: repeat-x;
  background-size: 5px 1px;
}

label ~ .c1 {
  margin-top: 16px;
}

.c1::before,
.c1::after {
  position: absolute;
  bottom: 0px;
  overflow: hidden;
  font-size: 0px;
}

<div
  className="c0 sinoui-base-input__layout c1 sinoui-input sinoui-input--disabled"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`渲染密集模式下的Input 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  line-height: 1.1875em;
}

.c0 > input,
.c0 > textarea {
  padding: 6px 0 7px;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.1876em;
  min-height: 1.1876em;
}

.c0 > input::-moz-placeholder,
.c0 > textarea::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder,
.c0 > textarea::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder,
.c0 > textarea:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear,
.c0 > textarea::-ms-clear {
  display: none;
}

.c0 > input:focus,
.c0 > textarea:focus {
  outline: 0;
}

.c0 > input:invalid,
.c0 > textarea:invalid {
  box-shadow: none;
}

.c0 > textarea {
  resize: none;
  overflow: hidden;
}

.c1 {
  color: rgba(0,0,0,0.87);
}

.c1::after {
  background-color: #3f51b5;
  left: 0;
  bottom: 0;
  content: '';
  height: 2px;
  position: absolute;
  right: 0;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  -webkit-transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c1::before {
  background-color: rgba(0,0,0,0.42);
  left: 0;
  bottom: 0;
  content: '';
  display: block;
  height: 1px;
  position: absolute;
  right: 0;
  -webkit-transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: background-color 200ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
}

.c1:hover::before {
  background-color: rgba(0,0,0,0.87);
  height: 2px;
}

.c1 > input {
  padding-top: 3px;
}

label ~ .c1 {
  margin-top: 16px;
}

.c1::before,
.c1::after {
  position: absolute;
  bottom: 0px;
  overflow: hidden;
  font-size: 0px;
}

<div
  className="c0 sinoui-base-input__layout c1 sinoui-input sinoui-input--dense"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;

exports[`渲染错误状态下的input 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: rgba(0,0,0,0.87);
  font-size: 1rem;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI','PingFang SC','Hiragino Sans GB','Microsoft YaHei',Roboto,'Noto Sans SC','Helvetica Neue',Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol';
  line-height: 1.1875em;
}

.c0 > input,
.c0 > textarea {
  padding: 6px 0 7px;
  display: block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  line-height: inherit;
  font: inherit;
  color: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  border: none;
  box-sizing: content-box;
  background: none;
  margin: 0px;
  -webkit-tap-highlight-color: transparent;
  min-width: 0px;
  width: 100%;
  height: 1.1876em;
  min-height: 1.1876em;
}

.c0 > input::-moz-placeholder,
.c0 > textarea::-moz-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-webkit-input-placeholder,
.c0 > textarea::-webkit-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input:-ms-input-placeholder,
.c0 > textarea:-ms-input-placeholder {
  color: rgba(0,0,0,0.38);
}

.c0 > input::-ms-clear,
.c0 > textarea::-ms-clear {
  display: none;
}

.c0 > input:focus,
.c0 > textarea:focus {
  outline: 0;
}

.c0 > input:invalid,
.c0 > textarea:invalid {
  box-shadow: none;
}

.c0 > textarea {
  resize: none;
  overflow: hidden;
}

.c1 {
  color: rgba(0,0,0,0.87);
}

.c1.c1::after {
  background-color: #f44336;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
}

.c1::after {
  background-color: #3f51b5;
  left: 0;
  bottom: 0;
  content: '';
  height: 2px;
  position: absolute;
  right: 0;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  -webkit-transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  transition: transform 200ms cubic-bezier(0.0,0.0,0.2,1) 0ms;
  pointer-events: none;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

label ~ .c1 {
  margin-top: 16px;
}

.c1::before,
.c1::after {
  position: absolute;
  bottom: 0px;
  overflow: hidden;
  font-size: 0px;
}

<div
  className="c0 sinoui-base-input__layout c1 sinoui-input sinoui-input--error"
  data-testid="baseInput"
  onClick={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
  />
</div>
`;
